File: C++/10.装饰器模式/DecoratorPatternDemo.cpp
/**
 * @file DecoratorPatternDemo.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-11-08
 *
 * @copyright Copyright (c) 2019
 *
 */

#include "Circle.h"
#include "Rectangle.h"
#include "RedShapeDecorator.h"

int main(int argc, char const *argv[]) {
    std::shared_ptr<Shape> circle = std::make_shared<Circle>();
    std::shared_ptr<ShapeDecorator> red_circle = std::make_shared<RedShapeDecorator>(std::make_shared<Circle>());
    std::shared_ptr<ShapeDecorator> red_rectangle = std::make_shared<RedShapeDecorator>(std::make_shared<Rectangle>());

    std::cout << "Circle with normal border" << std::endl;
    circle->Draw();

    std::cout << "\nCircle of red border" << std::endl;
    red_circle->Draw();

    std::cout << "\nRectangle of red border" << std::endl;
    red_rectangle->Draw();

    return 0;
}

File: Objective-C/10.DecoratorPattern/DecoratorPatternDemo.m
//
//  DecoratorPatternDemo.m
//  10.DecoratorPattern
//
//  Created by Layne on 2020/10/28.
//

#import "Circle.h"
#import "Rectangle.h"
#import "RedShapeDecorator.h"

int main(int argc, const char* argv[]) {
    @autoreleasepool {
        NSLog(@"Hello World!.");
        id<ShapeProtocol> circle = [[Circle alloc] init];
        ShapeDecorator* redCircle = [[RedShapeDecorator alloc] initWithShape:[[Circle alloc] init]];
        ShapeDecorator* redRectangle = [[RedShapeDecorator alloc] initWithShape:[[Rectangle alloc] init]];

        NSLog(@"Circle with normal border");
        [circle draw];

        NSLog(@"\nCircle of red border");
        [redCircle draw];

        NSLog(@"\nRectangle of red border");
        [redRectangle draw];
    }
    return 0;
}
